src/Data/Yaml.php

```diff
@@ -57,14 +57,11 @@ public static function decode($yaml): array
 
         // remove BOM
         $yaml   = str_replace("\xEF\xBB\xBF", '', $yaml);
-        $result = Spyc::YAMLLoadString($yaml);
+        $result = Parser::parse($yaml);
 
-        if (is_array($result)) {
-            return $result;
-        } else {
-            // apparently Spyc always returns an array, even for invalid YAML syntax
-            // so this Exception should currently never be thrown
-            throw new Exception('YAML string is invalid'); // @codeCoverageIgnore
-        }
+        // ensure that single string is wrapped in array
+        $result = A::wrap($result);
```

Comment:
I'm not sure if we should wrap the result if it's not already an array. While this can be convenient, it could also cause issues when the wrong kind of data is parsed by accident.

The Exception we had in the old implementation was there on purpose, it was just apparently never triggered because Spyc didn't allow it.

Reply:
@lukasbestle I put it in cause everything crashes when a single string is returned instead of an array with a single string inside. Spyc always returned arrays.

Reply:
Yes, but as I said: That's on purpose. The `Data` classes are always expected to deal with arrays, not scalar values. So if you pass a YAML string that only contains a string or number or whatever, that's not a valid input for the `Data` classes. Instead an Exception should be thrown like in the `JSON` handler.

Or is there code already that relies on this behavior that Spyc always returned an array?

Reply:
Yes, our code relies on that. If I put it in without the array wrapper, tests start failing left and right.

Reply:
Oh noes. That‘s bad. 😕

Reply:
I just tested it. Only three tests fail for the three field methods `toFile`, `toPage` and `toUser`. They fail because the current implementations of the respective plural methods support fields with a single string instead of an array of strings. We could fix those methods by checking if the field value begins with `-`. If it doesn't, there's only one name in the field and we can use it without trying to parse it as YAML.

Reply:
I'm not sure if we need to make it that complicated. The library is much more rigid with wrong data and actually throws exceptions itself - not like Spyc. And I feel these are much more helpful to use directly than covering up with any generic exception from us.

So it's really just about the case where a single string, int or so is returned instead of wrapped in an array.

Reply:
> So it's really just about the case where a single string, int or so is returned instead of wrapped in an array.

Yes, it is! That's what we are discussing about, aren't we?

The previous implementation has such a check that validates whether we got an array. If yes, the array is returned, otherwise an Exception is thrown. If there's a parsing error in the YAML input, the code won't even be run as the Exception that is thrown by Symfony YAML will break out of the `Yaml::decode()` method.

So we actually only need to revert these lines to the previous/current implementation and only need to remove the code coverage ignore marker and the comment.

Reply:
I read our conversation again but I still don't quite get your point.

SPYC would always return an array for valid YAML. 
So checking if the result is an array, if so returning it, if not throwing an Exception - that made sense.

The new library will return either an array or a single scalar for valid YAML.
If we keep the old code, we would throw an Exception for the latter cases, which would be a false negative.
The new library also throws reliably exceptions for invalid YAML (which SPYC didn't). So we can rely on this, I think.

Reply:
The point is: With SPYC, the exception was never thrown, not even for invalid YAML or any kind of input – SPYC *always* returns an array. However, the exception *should* have been thrown for non-array values and that's the reason why there currently is the comment above that tells the reader that this is a known issue.

The reason why non-arrays should trigger an Exception is the following: The `Handler::decode()` method, which `Yaml::decode()` inherits from, only allows array return values, never scalar values (because the `Data` classes are designed to always deal with array data).

So if a non-array value is parsed by the decoder, that should trigger a clear error as the user tried to parse data that the `Data` class is not designed to work with. If we don't throw an exception, the method will try to return a non-array value from a method that must only return arrays. This will trigger a PHP Fatal Error, which should be avoided.

I suggest the following implementation:

```php
public static function decode($yaml): array
{
    if ($yaml === null) {
        return [];
    }

    if (is_array($yaml) === true) {
        return $yaml;
    }

    $result = Parser::parse($yaml);

    if (is_array($result)) {
        return $result;
    } else {
        throw new Exception('YAML string is invalid');
    }
}
```

Again: It's *not* about YAML that is syntactically invalid. Those cases will throw an Exception inside Symfony Yaml, so the `if (is_array())` code won't even be run. This is just about the case of non-array YAML data. Also see the `Data\Json` class with the same implementation.

Reply:
But that’s why I put the A::wrap in there: we know that the YAML is valid because the new parser didn’t throw an exception. To comply with our return types, we wrap it in an array.

Reply:
Looks like we can discuss forever. We have exchanged those same arguments several times now. :)

I still don't think `A::wrap()` is a good idea because lax parser handling can cause unexpected issues down the road, but if you want to do it that way, sure.

Reply:
Yea I guess we have opposing opinions here.
I think it’s bad to throw an exception ourselves mentioning invalid YAML when the YAML is actually valid.